IA-2773: update payments api

quang-le opened the pull request:
Explain what problem this PR is resolving

Related JIRA tickets : IA-2773

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [X] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
in the code

## Changes

- Add an endpoint creating a `Task` for bulk update of payments. Django's `update()` wasn't used because we need to copy the original instance of  `Payment` to be able to log the changes in a `Modification`. This seemed the safest approach
- Add an endpoint to GET and PATCH `Payment`. The only updatable field is `status` for now
- Add `TaskAPITestCase` which extends `APITestCase` and add `Task` specific methods (taken from `org_units_bulk_update`)
- Connect front-end to both endpoints
- minor front-end improvements
- Add `AuditMixin` to the `Payment`and `PaymentLot` viewsets

## How to test

Go to Payments > Payment Lots
Select a payment lot with several payments
Try updating payments, single and in bulk

## Print screen / video

https://github.com/BLSQ/iaso/assets/38907762/d72dc9b8-4be5-4168-ac95-46528f0a9587


## Notes

The setup of the audit logs is not complete and will be handled via another PR, as well as updating the front-end to show when a payment is being updated by the worker.

quang-le committed:
IA-2773: add payments endpoint

- Add PaymentViewSet that accepts GET and PATCH
- All fields except "status" are read-only
- Set updated_by to request user in serializer
- Add tests

quang-le committed:
IA-2773: update front-end to use API endpoint

- usePATCH on /api/payments/

quang-le committed:
IA-2774: pass status to bulk update (front-end)

quang-le committed:
IA-2773: add bulk update endpoint

quang-le committed:
black

quang-le committed:
IA-2773: improve PaymentLot modal table loading state

- move all mutation hooks up to EditPaymentDialog
- pass mutation functions as props
- use isLoading from mutations to disable buttons and set tabels loading states

quang-le committed:
IA-2773: add tests

quang-le committed:
Merge branch 'IA-2774_add_payment_edition_modal' into IA-2773_update_payments_api

quang-le committed:
IA-2773: add translations

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/hooks/config/usePaymentColumns.tsx:
```diff
@@ -1,14 +1,18 @@
+    saveStatus,
 }: {
     potential?: boolean;
+    saveStatus?: UseMutateAsyncFunction<any, any, SavePaymentStatusArgs, any>;
```
❤️

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/hooks/requests/useSavePaymentStatus.ts:
```diff
@@ -2,42 +2,58 @@ import { UseMutationResult } from 'react-query';
+        select_all: body.selectAll,
+        status: body.status,
+    };
+    await waitFor(500);
```
why ?

quang-le committed:
Merge branch 'main' into IA-2773_update_payments_api

beygorghor reviewed (approved):
Working like a charm for me

quang-le committed:
IA-2773: fix typo

quang-le committed:
IA-2779: add AuditLogger class

quang-le committed:
IA-2779: add API labels to audit model

quang-le committed:
IA-2779: WIP: move bulk calculations to task (worker)

- create task for : payment lot creation, bulk payment update
- remove payment lot status calculation from model save method
- use AuditLogger to log updates
- known bugs: user is none on create_payments_from_payment_lot task

quang-le committed:
IA-2779: fix bugs

- fix wrong variable names and other syntax errors

quang-le committed:
black formatting

quang-le committed:
IA-2779: fix wrong translation key

quang-le committed:
Fix: remove camelcase check from es-lint

quang-le committed:
IA-2779: pass payment lot id when bulk updating payments

quang-le committed:
IA-2779: further update front end hooks

quang-le committed:
IA-2779: fix task bugs

- correctly log
- add missing return values

quang-le committed:
IA-2779: prevent dupliacte payments lots

- add foreign key from PotentialPAyemnt to PaymentLot
- filter out PotentialPayments associated to a PaymentLot in PotentialPaymentViewset, as this means there's a task running for that PaymentLot and these PotentialPayments
- add migration file

quang-le committed:
IA-2779: add task to payment lot

- add fk to model + migration
- add the task to the api payload sent to front-end

quang-le committed:
IA-2779: trim down payment lot response

quang-le committed:
IA-2779: handle payemnt lo task status in front-end

- Payment lot: add rowProps to tabel, so rows of payment lot with running or queued task will have different bacjground color and buttons will be disabled
- Payment lot modal:
    - Payment table is "loading" when associated payment lot has running or queued task
    - columns also have background color
    - all multi select buttons are disabled: select/undelect all + edit selected
- Update some TS types
- remove waitFor in patch and post request (dev leftovers)
- Add prop to disable EditIconButton

quang-le committed:
IA-2779: add translations

quang-le committed:
IA-2779: add permissions on logs API

quang-le committed:
black

quang-le committed:
IA-2779: fix existing tests

quang-le committed:
IA-2779: use JSON widget in Audit admin

quang-le committed:
IA-2779: Add tests for tasks

- payment lot creation
- mark all payemnt as read
- minor bug fixes

quang-le committed:
IA-2779: bug fix

- filter change requets on potential payment without payment lot

quang-le committed:
IA-2779: add missing test

quang-le committed:
IA-2779: add comments (doc)

quang-le committed:
IA-2779: add RefreshButton

quang-le committed:
IA-2779: add rRefreshButton to LotsPayments.tsx

mathvdh committed:
Merge pull request #1203 from BLSQ/IA-2779_improve_payment_audit

IA-2779: improve payment audit

mathvdh committed:
Merge branch 'main' of github.com:BLSQ/iaso into IA-2773_update_payments_api

mathvdh committed:
IA-2773 Fix migrations

mathvdh committed:
IA-2773 Fix tests
